adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1233,4 +1193,66 @@ private static void addRetainAnnotation(final MqttContext context, final Message
             MessageHelper.addAnnotation(downstreamMessage, MessageHelper.ANNOTATION_X_OPT_RETAIN, Boolean.TRUE);
         }
     }
+
+    private Future<TenantObject> isAdapterEnabled(final DeviceCredentials credentials, final TenantObject tenantConfig) {
+        if (tenantConfig.isAdapterEnabled(getTypeName())) {
+            LOG.debug("protocol adapter [{}] is enabled for tenant [{}]",
+                    getTypeName(), credentials.getTenantId());
+            return Future.succeededFuture(tenantConfig);
+        } else {
+            LOG.debug("protocol adapter [{}] is disabled for tenant [{}]",
+                    getTypeName(), credentials.getTenantId());
+            return Future.failedFuture(new ClientErrorException(HTTP_FORBIDDEN,
+                    "adapter disabled for tenant"));
+        }
+    }
+
+    private Future<DeviceUser> authenticateCredentials(final DeviceCredentials credentials) {
+        final Future<DeviceUser> result = Future.future();
+        usernamePasswordAuthProvider.authenticate(credentials, result.completer());
+        return result;
+    }
+
+    private Future<Device> isDeviceRegistered(final Device authenticatedDevice, final DeviceCredentials credentials,
+            final Span currentSpan) {
+        currentSpan.log(String.format("device authenticated"));
+        LOG.debug("successfully authenticated device [tenant-id: {}, auth-id: {}, device-id: {}]",
+                authenticatedDevice.getTenantId(), credentials.getAuthId(),
+                authenticatedDevice.getDeviceId());
+        return getRegistrationAssertion(authenticatedDevice.getTenantId(),
+                authenticatedDevice.getDeviceId(),
+                authenticatedDevice, currentSpan.context())
+                        .compose(ok -> accepted(authenticatedDevice));
+    }
+
+    private Future<Device> createLink(final MqttEndpoint endpoint, final Device authenticatedDevice,
+            final Span currentSpan) {
+        currentSpan.log(String.format("device registration verified"));
+        LOG.debug("verified device registration [tenant-id: {}, device-id: {}]",
```

Comment:
this log statement assumes that this method has been invoked *after* the device's registration has been verified. However, there is nothing that prevents this method from being invoked in a completely different context. IMHO not *every* few statements need to be refactored into a separate method ...

Reply:
I consolidated into one method `createLink`.